### PR TITLE
Issue 231: Negation nodes should only retract a token downstream upon…

### PR DESCRIPTION
**Pull request by WilliamParker**
… right-activation with matching elements if there were no matching elements previously

### CODE REVIEW

File: src/main/clojure/clara/rules/engine.cljc [hunk 1]
```diff
@@ -659,19 +665,32 @@
...
-
-                          :when (matches-some-facts? token
-                                                     elements
-                                                     join-filter-fn
-                                                     condition)]
-                      token)))
+    ;; Put the previous elements into a persistent collection since LocalMemory will return
+    ;; a mutable list here on the JVM, and we need to examine the previous elements on
+    ;; a per-token basis here.
+    (let [previous-elements (into [] (mem/get-elements memory node join-bindings))]
```
Comment (mrrodriguez) [inline src/main/clojure/clara/rules/engine.cljc]: Why not avoid this copying of data and instead just hold off on the `mem/add-elements!` until after you are done looking at the `previous-elements`?
You could pull the `for` out of the `retract-tokens` and instead bind to the results of the `for` in the same `let`-binding as the `previous-elements`.  So then the `mem/add-elements!` wouldn't happen until after.

The issue to me just appears to be the delay in "looking" at the `previous-elements`.

Comment (WilliamParker) [inline src/main/clojure/clara/rules/engine.cljc]: +1  I've updated the pull request.

**Comment by mrrodriguez** [comment]
+1

**Comment by rbrush** [comment]
Merged!

**Pull request merged**